cluster/clustered_data.go

```diff
@@ -0,0 +1,199 @@
+
+The key that we store the state under includes the epoch in it, e.g. `my-data-key-prefix-0000000234`. When data is
+loaded, the key with the highest epoch prefix is loaded.
+*/
+type ClusteredData struct {
+	lock           sync.Mutex
+	dataBucketName string
+	dataKeyPrefix  string
+	stateMachine   *StateUpdator
+	objStoreClient objstore.Client
+	epoch          uint64
+	opts           ClusteredDataOpts
+	readyState     clusteredDataState
+	stopping       atomic.Bool
+}
+
+type clusteredDataState int
+
+const (
+	clusteredDataStateReady = iota
+	clusteredDataStateLoaded
+	clusteredDataStateStopped
+)
+
+func NewClusteredData(stateMachineBucketName string, stateMachineKeyPrefix string, dataBucketName string,
+	dataKeyPrefix string, objStoreClient objstore.Client, opts ClusteredDataOpts) *ClusteredData {
+	return &ClusteredData{
+		objStoreClient: objStoreClient,
+		dataBucketName: dataBucketName,
+		dataKeyPrefix:  dataKeyPrefix,
+		stateMachine: NewStateUpdator(stateMachineBucketName, stateMachineKeyPrefix, objStoreClient,
+			StateUpdatorOpts{}),
+		opts:       opts,
+		readyState: clusteredDataStateReady,
+	}
+}
+
+type ClusteredDataOpts struct {
+	AvailabilityRetryInterval time.Duration
+	ObjStoreCallTimeout       time.Duration
+}
+
+func (mo *ClusteredDataOpts) setDefaults() {
+	if mo.AvailabilityRetryInterval == 0 {
+		mo.AvailabilityRetryInterval = DefaultAvailabilityRetryInterval
+	}
+}
+
+func (m *ClusteredData) LoadData() ([]byte, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.readyState == clusteredDataStateStopped {
+		return nil, errors.New("stopped")
+	}
+	// Atomically increment the epoch
+	buff, err := m.stateMachine.Update(func(state []byte) ([]byte, error) {
+		epoch := buffToEpoch(state)
+		newState := make([]byte, 8)
+		binary.BigEndian.PutUint64(newState, epoch+1)
+		return newState, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	m.epoch = buffToEpoch(buff)
+	prevEpoch := m.epoch - 1
+	// Now we try and load the key with the highest epoch - the key might be lower than prevEpoch as no data might
+	// have been stored in previous epoch
+	var data []byte
+	for {
+		dataKey := m.createDataKey(prevEpoch)
+		data, err = m.getWithRetry(dataKey)
+		if err != nil {
+			return nil, err
+		}
+		if len(data) == 0 {
+			// no key found - try with next lower epoch
+			if prevEpoch == 0 {
+				// No data to load
+				break
+			}
+			prevEpoch--
+		} else {
+			break
+		}
+	}
+	m.readyState = clusteredDataStateLoaded
+	return data, nil
+}
+
+func (m *ClusteredData) StoreData(data []byte) (bool, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.readyState != clusteredDataStateLoaded {
+		return false, errors.New("not loaded")
+	}
+	dataKey := m.createDataKey(m.epoch)
+	if err := m.putWithRetry(dataKey, data); err != nil {
+		return false, err
+	}
+	// Then we get latest epoch using a no-op update
+	buff, err := m.stateMachine.Update(func(buff []byte) ([]byte, error) {
```

Comment:
so to get it right, this will create a new key in S3 (PutIfNotExists) and we're using it only to:
- ensure that `key.notExists()` (so that we know we can get it from memory)
- or, if it does exist, fetch the latest state from S3 for that key?

Reply:
nit: I would abstract this use case into a method called `FetchLatest()` or something and write up the guarantees there too

Reply:
> so to get it right, this will create a new key in S3 (PutIfNotExists) and we're using it only to:

ensure that key.notExists() (so that we know we can get it from memory)
or, if it does exist, fetch the latest state from S3 for that key?

We're basically just checking if the epoch has been changed by another leader, and if so, fail the writer (protect against zombies).

Potentially can be replaced with a ListObjectsWithPrefix... ?